src/maximum_cardinality_matching.jl

```diff
@@ -0,0 +1,118 @@
+"""
+maximum_cardinality_matching(g::Graph)
+
+Given a graph `g` returns a maximum cardinality matching.
+This is the same as running `maximum_weight_matching` with default weights (`1`)
+but is faster without needing a JuMP solver.
+
+Returns MatchingResult containing:
+  - the maximum cardinality that can be achieved
+  - a list of each vertex's match (or -1 for unmatched vertices)
+"""
+function maximum_cardinality_matching(g::AbstractGraph{U}) where U<:Integer
+    n = nv(g)
+    matching = -ones(Int, n)
+    
+    # the number of edges that can possibly be part of a matching
+    max_generally_possible = fld(n, 2)
+
+    # get initial matching
+    matching_len = 0
+    for e in edges(g)
+        # if unmatched
+        if matching[e.src] == -1 && matching[e.dst] == -1
+            matching[e.src] = e.dst
+            matching[e.dst] = e.src
+            matching_len += 1
+        end
+    end
+
+    # if there are at least two free vertices
+    if matching_len < max_generally_possible
+
+        parents = zeros(Int, n)
+        visited = falses(n)
+        
+        @inbounds while matching_len < max_generally_possible
+            cur_level = Vector{Int}()
+            sizehint!(cur_level, n)
+            next_level = Vector{Int}()
+            sizehint!(next_level, n)
+
+            # get starting free vertex
+            free_vertex = 0
+            found = false
+            for v in vertices(g)
+                if matching[v] == -1
+                    visited[v] = true
+                    free_vertex = v
+                    push!(cur_level, v)
+           
+                    level = 1
+                    found = false
+                    # find augmenting path
+                    while !isempty(cur_level)
+                        for v in cur_level
+                            
+                            if level % 2 == 1
+                                for t in outneighbors(g, v)
+                                    # found an augmenting path if connected to a free vertex
+                                    if matching[t] == -1
+                                        # traverse the augmenting path backwards and change the matching
+                                        current_src = t
+                                        current_dst = v
+                                        back_level = 1
+                                        while current_dst != 0
+                                            # add every second edge to the matching (this also overwrites the current matching)
+                                            if back_level % 2 == 1
```

Comment:
Is level meant for deciding on the path if we go from an unmatched to a matched edge and vice-versa? In that case, I would suggest to use a boolean instead as division and modulo operations are quite slow.

Reply:
Good point yes a boolean variable should work as well. Thanks for your review :)
Nevertheless this PR isn't very helpful anymore, right? Or is there a need to have it for bipartite graphs as long as there is no implementation which works without the Blossom V binary?

Reply:
I would actually have to check, if we already have an algorithm for bipartite graphs. If not, we could use one for sure.

Otherwise, maybe this PR can be extended? In this case, we do not have any weights, so we don't actually need BlossomV, but the original blossom algorithm shoudl be sufficient, or maybe there is a newer better variant. Would be a good learning exercise, for implementing BlossomV later.

Also, the approach with just finding augmenting paths by bfs should work in general, it's just, that without special handling of blossoms, we will would have to do some backtracking so that we cannot guarantee polynomial running time anymore.

Reply:
Sounds like a good approach to me to extend this to general graphs using the standard blossom algorithm or research whether there are advancements to have polynomial running time.